spdx-utils/src/main/resources/declared-license-mapping.yml

```diff
@@ -369,6 +369,7 @@
 "Ruby's": Ruby
 "SIL OPEN FONT LICENSE Version 1.1": OFL-1.1
 "SIL Open Font License 1.1 (OFL-1.1)": OFL-1.1
+"SOLACE CORPORATION LICENSE AGREEMENT": LicenseRef-ort-Solace-Software-EULA
```

Comment:
For declared license mappings I believe it's better to not do any entry for license strings which do not have an associated ID instead of implicitly defining ORT license identifier for which users need to figure out semantics and implement the rest: E.g. provide the license text to the license ID.

I prefer not to do this change. IIUC we haven't done similar cchnages in the past, what's your rationale to start it now?

Reply:
> E.g. provide the license text to the license ID.

I was actually planning to look into adding license text for "LicenseRef-ort-" license, but this requires some changes to the way we currently import licenses.

> I prefer not to do this change. IIUC we haven't done similar cchnages in the past, what's your rationale to start it now?

We need to be able to somehow handle this license *now* in rules. And as you can see be searching for "LicenseRef-ort-" in `declared-license-mapping.yml` [we have done this before](https://github.com/oss-review-toolkit/ort/pull/3898/files#R164).

> One issue is that the user looses the option to map that string to a license ID of his choice.

Doesn't a curation's `declaredLicenseMapping` take precedence?

Reply:
> We need to be able to somehow handle this license _now_ in rules.

Does the need come from resolving rule violations within Bosch? If so I could share a different workflow to see if it is an alternative for you.

Reply:
> Does the need come from resolving rule violations within Bosch?

Yes.

> If so I could share a different workflow to see if it is an alternative for you.

While I'm certainly open to learn about that alternative, I'm still asking myself what's wrong with this change. Isn't it potentially valuable for any ORT user to be able to handle this license? And doesn't it thus make sense to generally give this license a name / id to operate on, instead of every user implementing an alternative workflow?

Reply:
> While I'm certainly open to learn about that alternative, I'm still asking myself what's wrong with this change. Isn't it potentially valuable for any ORT user to be able to handle this license? And doesn't it thus make sense to generally give this license a name / id to operate on, instead of every user implementing an alternative workflow?

The workflow we use is:
1. Define the license identifier by placing a file containing the license text in the custom license texts dir, see also [reporter-command](https://github.com/oss-review-toolkit/ort/blob/master/cli/src/main/kotlin/commands/ReporterCommand.kt#L121).
2. Define a curation with a [declared_license_mapping](https://github.com/oss-review-toolkit/ort/blob/master/model/src/main/kotlin/PackageCurationData.kt#L99) for the declared license string
3. Classify that license identifier

Out of the critical path: 
1. Contribute the detection of these custom license ids to ScanCode
2. Remove (temporary) custom license IDs again, as soon as ORT detects them (via ScanCode).

The problem of the use case related to this PR i believe is that one cannot instantly address the issue properly, because there is no license identifier defined, neither in SPDX nor in ScanCode. The custom license IDs were invented in order to make the issue fixable instantly and enabling the long term ideal solution to be done outside the critical path (of compliance reviews). Implementing these temporary work-arounds in ORT IMO make things more complicated, amongst other since the license identifers semantic is only defined implicitly and not trivial to figure out. It at least needs investigating the git history. Further it adds yet an alternative way to adress the same thing, so one has to choose from two options now which makes things IMO a bit more complicate. Finally, I consider these license identifers ideally temporary (until e.g. ScanCode has an identifier for the respective license) - but license identifiers IMO just cannot be deletet but should be maintained as deprecated.

From the above I believe we shouldn't add license identifiers to ORT for this use case due to added complexity.

Reply:
> The custom license IDs were invented in order to make the issue fixable instantly

But ScanCode would do exactly the same! If ScanCode implements https://github.com/nexB/scancode-toolkit/issues/2498, they would also need to come up with a LicenseRef for the license, and they'd probably call it `LicenseRef-scancode-Solace-Software-EULA`. So why is it acceptable to use `LicenseRef-scancode-Solace-Software-EULA` in a mapping, but not `LicenseRef-ort-Solace-Software-EULA`?

> Finally, I consider these license identifers ideally temporary (until e.g. ScanCode has an identifier for the respective license)

Actually, I'm thinking about more or less the opposite: In the mid-term, we should stop emitting ScanCode-specific LicenseRefs from ORT, in favor of emitting ORT-specific LicenseRefs. Because I'm looking at adding more (relevant) scanner support soon, and with that in place, it might be that we (for some type of scans) do not use ScanCode anymore at all. In such a workflow it would be highly confusing to still emit ScanCode-specific LicenseRefs for declared licenses.

So, in the same way as ORT's scanner module wraps the details of running the different scanners and parsing their results, we should also wrap scanner-specific LicenseRefs IMO.

Reply:
> But ScanCode would do exactly the same! If ScanCode implements [nexB/scancode-toolkit#2498](https://github.com/nexB/scancode-toolkit/issues/2498), they would also need to come up with a LicenseRef for the license, and they'd probably call it `LicenseRef-scancode-Solace-Software-EULA`. So why is it acceptable to use `LicenseRef-scancode-Solace-Software-EULA` in a mapping, but not `LicenseRef-ort-Solace-Software-EULA`?

I believe that "ScanCode would do exactly the same" is not exactly right, as for scancode license identifiers it is straight forward to figure out the semantics and the corresponding licenses texts.  

> Actually, I'm thinking about more or less the opposite: In the mid-term, [...]

Ok, sure - if you consider wrapping license identifiers as a mid (or long term?) goal, that'd change the basis of the discussion. However as long as there is no agreement on that I wouldn't assume it as a basis for deciding whether this PR makes sense for the current design.

Reply:
Anyhow, I'd not block this change - but be interested in further opinions. maybe @oss-review-toolkit/core-devs anyone?

Reply:
I agree with @fviernau's point that defining ORT specific LicenseRefs without providing additional information is a bad practice. We should provide more information, e.g. include the license text or information how to find the license text and additional information about the license.
But because we already have ORT specific LicenseRefs in this file, and because I believe we should discuss this topic separately to agree on a solution without blocking this PR, I will approve it.

Reply:
Thanks. I've created https://github.com/oss-review-toolkit/ort/issues/3910 as a follow-up.